backend/de.metas.business/src/main/java/de/metas/payment/C_Payment.java

```diff
@@ -102,7 +102,12 @@ public void updateFromOrder(@NonNull final I_C_Payment record)
 		final BigDecimal priceActual = paymentTermDiscountPercent.subtractFromBase(order.getGrandTotal(), currency.getPrecision().toInt());
 		final BigDecimal discountAmount = paymentTermDiscountPercent.computePercentageOf(order.getGrandTotal(), currency.getPrecision().toInt());
 
-		record.setC_BPartner_ID(firstGreaterThanZero(order.getBill_BPartner_ID(), order.getC_BPartner_ID()));
+		final BPartnerLocationId orderBPartnerLocationId = CoalesceUtil.coalesceNotNull(
+				BPartnerLocationId.ofRepoIdOrNull(order.getBill_BPartner_ID(), order.getBill_Location_ID()),
+				BPartnerLocationId.ofRepoIdOrNull(order.getC_BPartner_ID(), order.getC_BPartner_Location_ID()));
+
+		record.setC_BPartner_ID(orderBPartnerLocationId.getBpartnerId().getRepoId());
+		record.setC_BPartner_Location_ID(orderBPartnerLocationId.getRepoId());
```

Comment:
that covers the prepayment case.

not sure if we shall cover the case when we pay an invoice, i.e. user is setting the C_Invoice_ID (and not the C_Order).

More, usually the payments will be generated from bank statements, and them somebody will match them against invoices. Shall also cover something about this?

Reply:
Yep, thought about that, but it will increase the effort (especially for testing) and it wasn't really a requirement at this point... I'll check with Torby and see how we shall proceed. Main focus for abeam are the prepayments.